libcaf_net/src/application.cpp

```diff
@@ -0,0 +1,287 @@
+  if (auto err = source(src_node, src_id, dst_id, fwd_stack, content))
+    return err;
+  // Sanity checks.
+  if (dst_id == 0)
+    return ec::invalid_payload;
+  // Try to fetch the receiver.
+  auto dst_hdl = system().registry().get(dst_id);
+  if (dst_hdl == nullptr) {
+    CAF_LOG_DEBUG("no actor found for given ID, drop message");
+    return caf::none;
+  }
+  // Try to fetch the sender.
+  strong_actor_ptr src_hdl;
+  if (src_node != none && src_id != 0)
+    src_hdl = proxies_->get_or_put(src_node, src_id);
+  // Ship the message.
+  auto ptr = make_mailbox_element(std::move(src_hdl),
+                                  make_message_id(hdr.operation_data),
+                                  std::move(fwd_stack), std::move(content));
+  dst_hdl->get()->enqueue(std::move(ptr), nullptr);
+  return none;
+}
+
+error application::handle_resolve_request(write_packet_callback& write_packet,
+                                          header hdr, byte_span payload) {
+  CAF_ASSERT(hdr.type == message_type::resolve_request);
+  size_t path_size = 0;
+  binary_deserializer source{system(), payload};
+  if (auto err = source.begin_sequence(path_size))
+    return err;
+  // We expect the payload to consist only of the path.
+  if (path_size != source.remaining())
+    return ec::invalid_payload;
+  auto remainder = source.remainder();
+  string_view path{reinterpret_cast<const char*>(remainder.data()),
+                   remainder.size()};
+  // Write result.
+  auto result = resolve_local_path(path);
+  buf_.clear();
+  actor_id aid = result ? result->id() : 0;
+  std::set<std::string> ifs;
+  // TODO: figure out how to obtain messaging interface.
+  serializer_impl<buffer_type> sink{system(), buf_};
+  if (auto err = sink(aid, ifs))
+    return err;
+  auto out_hdr = to_bytes(header{message_type::resolve_response,
+                                 static_cast<uint32_t>(buf_.size()),
+                                 hdr.operation_data});
+  return write_packet(out_hdr, buf_);
+}
+
+error application::handle_resolve_response(write_packet_callback&, header hdr,
+                                           byte_span payload) {
+  CAF_ASSERT(hdr.type == message_type::resolve_response);
+  auto i = pending_resolves_.find(hdr.operation_data);
+  if (i == pending_resolves_.end()) {
+    CAF_LOG_ERROR("received unknown ID in resolve_response message");
+    return none;
+  }
+  auto guard = detail::make_scope_guard([&] {
+    if (i->second.pending())
+      i->second.deliver(sec::remote_lookup_failed);
+    pending_resolves_.erase(i);
+  });
+  actor_id aid;
+  std::set<std::string> ifs;
+  binary_deserializer source{system(), payload};
+  if (auto err = source(aid, ifs))
+    return err;
+  if (aid == 0) {
+    i->second.deliver(strong_actor_ptr{nullptr}, std::move(ifs));
+    return none;
+  }
+  i->second.deliver(proxies_->get_or_put(peer_id_, aid), std::move(ifs));
+  return none;
+}
+
+error application::generate_handshake() {
```

Comment:
Not sure if we ever want to retransmit handshakes. If so, it might be a good idea to clear `buf_` before filling it.